docs/multiple dataplane.md

```diff
@@ -0,0 +1,166 @@
+		- 新启动的Pod通过Kmesh编排寻址并短接envoy访问server端
+		- 存量Pod不受影响，通过envoy进行编排访问server端
+	- 用户指定namespace不再使用Kmesh作为数据面，此namespace、访问的svc中对应的pod中都安装了envoy
+		- 新启动Pod通过envoy编排后访问server端
+		- 存量Pod不受影响，Kmesh纳管的容器，仍通过Kmesh编排寻址并短接envoy来访问server端
+
+![](./pics/client_with_envoy.png)
+
+## 用户场景二：Kmesh纳管pod与sidecar(envoy)数据面中的pod进行通讯分析：
+
+场景描述
+
+- 纳管Pod上没有envoy
+	- 用户指定namespace使用Kmesh作为数据面，此namespace中未安装envoy，访问的svc中pod安装了envoy
+		- 新启动的Pod通过Kmesh编排寻址访问server端，server由envoy进行接收纳管
+		- 存量Pod不受影响，通过k8s原生svc进行访问server端，server由envoy进行接收纳管
+	- 用户指定namespace不再使用Kmesh作为数据面，此namespace中未安装envoy，访问的svc中pod安装了envoy
+		- 新启动Pod通过k8s原生svc进行访问server端，server由envoy进行接收纳管
+		- 存量Pod不受影响，Kmesh纳管的容器，仍通过Kmesh编排后访问server端，server由envoy进行接收纳管
+
+![](./pics/client_with_noenvoy_server_with_envoy.png)
+
+- 纳管Pod上有envoy
+	- 用户指定namespace使用Kmesh作为数据面，此namespace、访问的svc中对应的pod中都安装了envoy
+		- 新启动的Pod通过Kmesh编排寻址并短接envoy访问server端，server由envoy进行接收纳管
+		- 存量Pod不受影响，通过envoy进行编排访问server端，server由envoy进行接收纳管
+	- 用户指定namespace不再使用Kmesh作为数据面，此namespace、访问的svc中对应的pod中都安装了envoy
+		- 新启动Pod通过envoy编排后访问server端，server由envoy进行接收纳管
+		- 存量Pod不受影响，Kmesh纳管的容器，仍通过Kmesh编排寻址并短接envoy来访问server端，server由envoy进行接收纳管
+
+![](./pics/client_with_envoy_server_with_envoy.png)
+
+## usecase
+### 使用接口设计
+用户使用接口
+
+	# 对指定的namespace来启用Kmesh
+	kubectl label namespace xxx label istio.io/dataplane-mode=Kmesh
+	# 对指定的namespace来关闭Kmesh
+	kubectl label namespace xxx label istio.io/dataplane-mode-
+
+## 功能实现原理：
+### 组件设计
+要实现上述功能，从用户指定对特定namespace启用Kmesh到整体功能生效，涉及的组件如下图
+
+![](./pics/multiple_dataplane_design.png)
+
+Kmesh中需要进行修改的有如下组件：
+
+#### daemon
+
+daemon需要对cni插件进行管理，将Kmesh启动、cni pod重启时，将cni插件调用信息写入到/etc/cni/net.d的对应conflist中。calico的conflist的格式如下：
+
+	{
+		"name": "k8s-pod-network",
+		"cniVersion": "0.3.1",
+		"plugins": [
+			{
+				"type": "calico",
+				"log_level": "info",
+				...
+			},
+			{
+				"type": "portmap",
+				...
+			},
+			...
+			// 新增kmesh的cni-plugin
+			{
+				"type": "Kmesh-cniplugin",
+			}
+		]
+	}
+
+flannel的conflist格式如下：
+
+	{
+		"name": "cbr0",
+		"cniVersion": "0.3.1",
+		"plugins": [
+			{
+				"type": "flannel",
+				...
+			},
+			...
+			// 新增kmesh的cni-plugin
+			{
+				"type": "Kmesh-cniplugin",
+			}
+		]
+	}
+
+daemon变更如下：
+
+- 在kmesh使能时写入kemsh plugin配置到cni conflist中
+
+	在kmesh-daemon启动时，自动将cni plugin配置`{"type":"kmesh-cni"}`写入到/etc/cni/net.d/目录中的配置文件(.conflist结尾)中去
+
+- 在kmesh除能时从cni conflist中清理kmesh plugin配置 
+
+	在kmesh-daemon退出时，自动将cni plugin配置`{"type":"kmesh-cni"}`从/etc/cni/net.d/目录中的配置文件(.conflist结尾)中删除
+
+#### cni-plugin
+
+cni用于在集群创建新的Pod时，判断该Pod是否属于打上Kmesh标签的namespace。
+
+- 无`istio.io/dataplane-mode=Kmesh`，什么也不做
+- 有`istio.io/dataplane-mode=Kmesh`，则在新创建的Pod的cgroup中修改net_cls.classid为0x1000，**当前仅支持cgroupfs，不支持systemd**，cgroupfs的路径默认如下：/sys/fs/cgroup/net_cls/kubepods/podxxxxxx/net_cls.classid
+- 有`istio.io/dataplane-mode=Kmesh`且有`istio-injection=enabled`，则会进入到新建的Pod中加入以下iptables规则
+
+收包路径上所有数据包短接envoy:
+
+	iptables -t nat -I 1 PREROUTING -j RETURN

```

Comment:
I am not sure in understand this rule, as said above, the kmesh won't influence pod that has sidecar.

Same for below outbound rule

Reply:
Only the new start pod is affected. The iptables rule is injected to the new start pod with sidecar, and no operation is performed on the existing pod.